src/nccl_ofi_ep_addr_list.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All rights reserved.
+ */
+
+#include "config.h"
+
+#include "nccl_ofi.h"
+
+#include "nccl_ofi_ep_addr_list.h"
+#include "nccl_ofi_pthread.h"
+
+#include "contrib/uthash.h"
+#include "contrib/utlist.h"
+
+/**
+ * A Libfabric address, stored in a form hashable by uthash
+ */
+typedef struct {
+	char addr[MAX_EP_ADDR];
+	UT_hash_handle hh;
+} hashed_addr_t;
+
+/**
+ * A struct storing a pair of (ep, addr_set)
+ */
+typedef struct {
+	nccl_net_ofi_ep_t *ep;
+	/* Hash set of addresses the endpoint is connected to */
+	hashed_addr_t *addr_set;
+} pair_ep_addr_set_t;
+
+/**
+ * A linked list of pairs of (ep, HashSet<addr>). The list should be stored in
+ * the calling code as a pointer to this struct.
+ */
+typedef struct ep_pair_list_elem {
+	pair_ep_addr_set_t pair;
+	struct ep_pair_list_elem *prev;
+	struct ep_pair_list_elem *next;
+} ep_pair_list_elem_t;
+
+/**
+ * Outer structure storing the ep list and a mutex to protect access
+ */
+struct nccl_ofi_ep_addr_list {
+	ep_pair_list_elem_t *ep_pair_list;
+	pthread_mutex_t mutex;
+};
+
+void nccl_ofi_init_ep_addr_list(nccl_ofi_ep_addr_list_t **list)
+{
+	nccl_ofi_ep_addr_list_t *ret_list = calloc(1, sizeof(*ret_list));
+	if (!ret_list) {
+		NCCL_OFI_WARN("Failed to allocate list");
+		goto error;
+	}
+
+	ret_list->ep_pair_list = NULL;
+
+	if (nccl_net_ofi_mutex_init(&ret_list->mutex, NULL) != 0) {
+		NCCL_OFI_WARN("Failed to init mutex");
+		goto error;
+	}
+
+	goto exit;
+
+error:
+	if (ret_list) free(ret_list);
+	ret_list = NULL;
+
+exit:
+	*list = ret_list;
+}
+
+nccl_net_ofi_ep_t *nccl_ofi_get_ep_for_addr(nccl_ofi_ep_addr_list_t *ep_list, void *addr)
+{
+	nccl_net_ofi_mutex_lock(&ep_list->mutex);
+
+	ep_pair_list_elem_t *ep_pair = NULL;
+
+	nccl_net_ofi_ep_t *ret_ep = NULL;
+
+	DL_FOREACH(ep_list->ep_pair_list, ep_pair) {
+		hashed_addr_t *found_handle;
+		HASH_FIND(hh, ep_pair->pair.addr_set, (char *)addr, MAX_EP_ADDR, found_handle);
```

Comment:
This doesn't make sense.  There's no fixed size of an address, and you definitely can't assume it's MAX_EP_ADDR, so you're really just hashing junk here.  I'm really surprised this works; you're getting lucky by some calloc somewhere.

Reply:
This was intentional. We call `fi_getname` with output to `ep_rail`, which is a member of `ep`, which is calloc'd. So all our addresses are zero-padded up to `MAX_EP_ADDR`.

Passing around the lengths returned by `fi_getname` would be a lot of extra legwork. We'd have to store it in the rail info, the connect message, and with each address in ep_addr_list.

Alternatively, we could just manually zero-out the extra bits immediately after calling `fi_getname` to make it more obvious that the address is zero-padded?

Reply:
do it right; copy around the length.

Reply:
Done